hw16-serialization

SGrechnev opened the pull request.

SGrechnev committed:
hw16-serialization

stvort commented on hw16-serialization/src/main/java/ru/otus/GsonDemo.java:
```diff
@@ -0,0 +1,63 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class GsonDemo {
```
Давайте сделаем так, чтобы проходил следующий тест:
```
class MyGsonTest {

    @ParameterizedTest
    @MethodSource("generateDataForCustomTest")
    void customTest(Object o){
        Gson gson = new GsonBuilder().serializeNulls().create();
        MyGson myGson = new MyGson();
        assertThat(myGson.toJson(o)).isEqualTo(gson.toJson(o));
    }

    private static Stream<Arguments> generateDataForCustomTest() {
        return Stream.of(
                null,
                Arguments.of(true), Arguments.of(false),
                Arguments.of((byte)1), Arguments.of((short)2f),
                Arguments.of(3), Arguments.of(4L), Arguments.of(5f), Arguments.of(6d),
                Arguments.of("aaa"), Arguments.of('b'),
                Arguments.of(new byte[] {1, 2, 3}),
                Arguments.of(new short[] {4, 5, 6}),
                Arguments.of(new int[] {7, 8, 9}),
                Arguments.of(new float[] {10f, 11f, 12f}),
                Arguments.of(new double[] {13d, 14d, 15d}),
                Arguments.of(List.of(16, 17, 18)),
                Arguments.of(Collections.singletonList(19))
        );
    }
}
```
Необходимые зависимости:
```
    testImplementation 'org.junit.jupiter:junit-jupiter:5.7.0'
    testImplementation 'org.junit.jupiter:junit-jupiter-params:5.7.0'
    testImplementation 'org.assertj:assertj-core:3.18.1'
```

SGrechnev replied:
исправил, тест проходит + дополнил тест

stvort commented on hw16-serialization/src/main/java/ru/otus/gson/MyGson.java:
```diff
@@ -0,0 +1,92 @@
+    }
+
+    private static boolean isIntegerNumber(Field field) {
+        return Set.of(Byte.class, Byte.TYPE, Long.class, Long.TYPE, Short.class, Short.TYPE, Integer.class, Integer.TYPE)
+                .contains(field.getType());
+    }
+
+    private static boolean isFloatNumber(Field field) {
+        return Set.of(Double.TYPE, Double.class, Float.TYPE, Float.class).contains(field.getType());
+    }
+
+    private static boolean isBoolean(Field field) {
+        return (field.getType().equals(Boolean.TYPE));
+    }
+
+    private static boolean isString(Field field) {
+        return Set.of(String.class, Character.TYPE).contains(field.getType());
+    }
```
Возможно сеты стоит создать один раз

stvort commented on hw16-serialization/src/main/java/ru/otus/gson/MyGson.java:
```diff
@@ -0,0 +1,92 @@
+        return (field.getType().isArray());
+    }
+
+    private static List<Object> getListFromArray(Object array) {
+        if (array.getClass().isArray()) {
+            List<Object> objectList = new ArrayList<>();
+            for (int i = 0; i < Array.getLength(array); i++) {
+                objectList.add(Array.get(array, i));
+            }
+            return objectList;
+        }
+        return null;
+    }
```
А разве отсюда не проще вернуть сразу `JsonArray`?

stvort commented on hw16-serialization/src/main/java/ru/otus/gson/MyGson.java:
```diff
@@ -0,0 +1,92 @@
+        var builder = Json.createObjectBuilder();
+
+        for (Field field : obj.getClass().getDeclaredFields()) {
+            if (Modifier.isTransient(field.getModifiers())) {
```
Круто! Обычные константы тоже можно не сериализовать)

SGrechnev committed:
Add MyGsonTest
Fix MyGson for work with primitive types, arrays and collections
Add org.slf4j.Logger

SGrechnev committed:
Do not serialize static fields
